File: src/main/java/wraith/fabricaeexnihilo/api/registry/ICrucibleHeatRegistry.java
package wraith.fabricaeexnihilo.api.registry;

import net.minecraft.block.Block;
import net.minecraft.fluid.Fluid;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.tag.Tag;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.api.crafting.ItemIngredient;
import wraith.fabricaeexnihilo.api.recipes.crucible.CrucibleHeatRecipe;

import java.util.Collection;

public interface ICrucibleHeatRegistry extends IRegistry<CrucibleHeatRecipe> {

    default boolean register(ItemIngredient blocks, FluidIngredient fluids, int heat) {
        return register(new CrucibleHeatRecipe(blocks, fluids, heat));
    }

    default boolean register(Tag.Identified<Fluid> tag, int heat) {
        return register(ItemIngredient.EMPTY, new FluidIngredient(tag), heat);
    }

    default boolean register(Block block, int heat) {
        return register(new ItemIngredient(block), FluidIngredient.EMPTY, heat);
    }

    default boolean register(Fluid fluid, int heat) {
        return register(ItemIngredient.EMPTY, new FluidIngredient(fluid), heat);
    }

    int getHeat(Block block);

    int getHeat(Fluid fluid);

    int getHeat(Item item);

    int getHeat(ItemStack stack);

    // All recipes, chunked/broken up for pagination
    Collection<CrucibleHeatRecipe> getREIRecipes();

}

File: src/main/kotlin/exnihilofabrico/api/registry/ICrucibleHeatRegistry.kt
package exnihilofabrico.api.registry

import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.recipes.CrucibleHeatRecipe
import net.minecraft.block.Block
import net.minecraft.fluid.Fluid
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.tag.Tag

interface ICrucibleHeatRegistry {
    fun clear()
    fun register(recipe: CrucibleHeatRecipe): Boolean

    fun register(blocks: ItemIngredient, fluids: FluidIngredient, heat: Int) = register(CrucibleHeatRecipe(blocks, fluids, heat))

    fun registerFluidTag(tag: Tag<Fluid>, heat: Int) = register(ItemIngredient.EMPTY, FluidIngredient(tag), heat)

    fun register(block: Block, heat: Int) = register(ItemIngredient(block), FluidIngredient.EMPTY, heat)
    fun register(fluid: Fluid, heat: Int) = register(ItemIngredient.EMPTY, FluidIngredient(fluid), heat)

    fun getHeat(block: Block): Int
    fun getHeat(fluid: Fluid): Int
    fun getHeat(item: Item): Int
    fun getHeat(stack: ItemStack): Int
}